loading: work on simplifications (and some corrections) of docs (#29946)
JuliaLang/julia: doc/src/manual/code-loading.md

This manifest file describes a possible complete dependency graph for the `App` project:

> - There are two different packages named `Priv` that the application uses. It uses a private package, which is a root dependency, and a public one, which is an indirect dependency through `Pub`. These are differentiated by the uuid, and have different deps:

Reviewer: ```suggestion
- There are two different packages named `Priv` that the application uses. It uses a private package, which is a root dependency, and a public one, which is an indirect dependency through `Pub`. These are differentiated by their distinct UUIDs, and they have different deps:
```

What happens if `import Zebra` is evaluated in the main `App` code base? Since `Zebra` does not appear in the project file, the import will fail even though `Zebra` *does* appear in the manifest file. Moreover, if `import Zebra` occurs in the public `Priv` package—the one with UUID `2d15fe94-a1f7-436c-a4d8-07a9a496e01c`—then that would also fail since that `Priv` package has no declared dependencies in the manifest file and therefore cannot load any packages. The `Zebra` package can only be loaded by packages for which it appear as an explicit dependency in the manifest file: the  `Pub` package and one of the `Priv` packages.

> **The paths map** of a project environment is also determined by searching through the manifest files that are present in the load path. The path to the folder for a given package, identified by `uuid` and `Name`, is determined by these rules (in order):

Reviewer: ```suggestion
**The paths map** of a project environment is also determined by the manifest file. If there is no manifest file, then the paths map of the environment is empty. The path of a package `uuid` named `X` is determined by these two rules:
```

Author: > If there is no manifest file, then the paths map of the environment is empty

If there's no manifest (or an empty manifest), the path map may still contain a single entry (the path to the project environment itself)

> 

Reviewer: I'm fairly certain that the changes here are wrong.

Author: What part?

> 1. If the directory contains a Project file with a matching `uuid` and `Name`, then either:

Reviewer: ```suggestion
1. If the directory contains a project file with a matching `uuid` and `name`, then either:
```

>   - It has a `path` entry, which will be used (relative to the Project file directory).

Reviewer: ```suggestion
  - It has a `path` entry, that will be used (interpreted relative to the directory containing the project file).
```

>   - Otherwise, it'll use the Project file directory.

Reviewer: This is not correct. I'm not sure where you're getting this.

Author: Consider `Pkg.activate("stdlib/Pkg-f9180e48b27a843aeee864db814dce57cb296b9b")`. This doesn't have a Manifest, but the path map still has one entry (the Project file directory itself, as seen by doing `using Pkg`). That's the case that the additional rule added here covers.

If, on the other hand, Julia was loading the *other* `Priv` package—the one with UUID `2d15fe94-a1f7-436c-a4d8-07a9a496e01c`—it finds its stanza in the manifest, see that it does *not* have a `path` entry, but that it does have a `git-tree-sha1` entry. It then computes the `slug` for this UUID/SHA-1 pair, which is `HDkrT` (the exact details of this computation aren't important, but it is consistent and deterministic). This means that the path to this `Priv` package will be `packages/Priv/HDkrT/src/Priv.jl` in one of the package depots. Suppose the contents of `DEPOT_PATH` is `["/home/me/.julia", "/usr/local/julia"]`, then Julia will look at the following paths to see if they exist:

1. `/home/me/.julia/packages/Priv/HDkrT`
2. `/usr/local/julia/packages/Priv/HDkrT`

> Julia uses the first of these that exists to try to load the public `Priv` package, either from the file `HDkrT`, or the file `HDKrT/src/Priv.jl`.

Reviewer: I don't recall that we support `HDkrT` being a file and loading it as Julia source. Are you sure about this?

Author: This is just documenting more accurately what the code actually does. I do agree it's an odd thing to support and that we should perhaps consider dropping this functionality.

Reviewer: We should definitely remove that feature rather than document it. Let's do so before anyone decides to rely on it.

Reviewer: ```suggestion
Julia uses the first of these that exists to try to load the public `Priv` package from the file `packages/Priv/HDKrT/src/Priv.jl` in the depot where it was found.
```

What happens if `import Zebra` is evaluated in the main `App` code base? Since `Zebra` does not appear in the project file, the import will fail even though `Zebra` *does* appear in the manifest file. Moreover, if `import Zebra` occurs in the public `Priv` package—the one with UUID `2d15fe94-a1f7-436c-a4d8-07a9a496e01c`—then that would also fail since that `Priv` package has no declared dependencies in the manifest file and therefore cannot load any packages. The `Zebra` package can only be loaded by packages for which it appear as an explicit dependency in the manifest file: the  `Pub` package and one of the `Priv` packages.

**The paths map** of a project environment is extracted from the manifest file. The path of a package `uuid` named `X` is determined by these rules (in order):

1. If the project file in the directory matches `uuid` and name `X`, then either:

>   - It has a toplevel `path` entry, that will be used (interpreted relative to the directory containing the project file).

Reviewer: ```suggestion
  - It has a toplevel `path` entry, then (`uuid`, `X`) will be mapped to that path, interpreted relative to the directory containing the project file.
```

>   - Otherwise, it'll use the Project file directory.

Reviewer: ```suggestion
  - Otherwise, (`uuid, `X`) is mapped to  `src/X.jl` relative to the directory containing the project file.
```

> 2. If the project file has a corresponding manifest file, and the manifest contains a stanza matching `uuid`:

Reviewer: ```suggestion
2. If the above is not the case and the project file has a corresponding manifest file and the manifest contains a stanza matching `uuid` then:
```

>   - If it has a `path` entry, use that path (relative to the manifest file).

Reviewer: ```suggestion
  - If it has a `path` entry, use that path (relative to the directory containing the manifest file).
```

  - If it has a `git-tree-sha1` entry, compute a deterministic hash function of `uuid` and `git-tree-sha1`—call it `slug`—and look for a directory named `packages/X/$slug` in each directory in the Julia `DEPOT_PATH` global array. Use the first such directory that exists.

>   - Otherwise, ignore this stanza.

Reviewer: ```suggestion
  - Otherwise, there is no path mapping for (`uuid`, `X`).
```

1. `/home/me/.julia/packages/Priv/HDkrT`
2. `/usr/local/julia/packages/Priv/HDkrT`

> Julia uses the first of these that exists to try to load the public `Priv` package from the file `packages/Priv/HDKrT/src/Priv.jl` in the depot where it was found.

Author: OR the file `packages/Priv/HDKrT`—or is the intent to correct the code to match?

)
```

This example map includes three different kinds of package locations (the first and third are part of the default load path):

1. The private `Priv` package is "[vendored](https://stackoverflow.com/a/35109534/659248)" inside the `App` repository.
2. The public `Priv` and `Zebra` packages are in the system depot, where packages installed and managed by the system administrator live. These are available to all users on the system.
3. The `Pub` package is in the user depot, where packages installed by the user live. These are only available to the user who installed them.

### Package directories

> Package directories provide a simpler kind of environment without the ability to handle name collisions. In a package directory, the set of top-level packages is the set of subdirectories that "look like" packages. A package `X` is exists in a package directory if the directory contains one of the following "entry point" files:

Reviewer: If you're just in a package directory environment, you cannot handle package name collisions at all. In a stacked environment where some of the things in the stack are package directories name collisions are fine, but I don't think that's relevant to this description. I'm guessing that "less careful reproducibility of versions" is referring to the fact that one can't snapshot the exact set of versions in a project file, but I feel like that's incidental: if you check the full contents of a package directory into git and version control it, then it's equally reproducible. So I stand by my wording of this.

)
```

This example map includes three different kinds of package locations (the first and third are part of the default load path):

1. The private `Priv` package is "[vendored](https://stackoverflow.com/a/35109534/659248)" inside the `App` repository.
2. The public `Priv` and `Zebra` packages are in the system depot, where packages installed and managed by the system administrator live. These are available to all users on the system.
3. The `Pub` package is in the user depot, where packages installed by the user live. These are only available to the user who installed them.

### Package directories

> Package directories provide a simpler environment, but less powerful or controlled, and they exist in two flavors (depending on whether they contain a project file). The set of root packages available is defined simply as the set of files in the directory that "look like" packages. A package `Name` is determined to exist if it finds one of the following files:

Reviewer: This is not correct: if there's project file then it's always interpreted as a project environment, never as a package directory—that's the litmus test for which one you have.

Reviewer: ```suggestion
Package directories provide a simpler kind of environment with less control and without the ability to handle name collisions. In a package directory, the set of root packages available is just determined by the subdirectories that "look like" packages. A package `X` is determined to exist in a package directory if it contains one of the following "entry point" files:
```

  - A file `Name.jl`
  - A file `Name/src/Name.jl`
  - A file `Name.jl/src/Name.jl`

> Additionally, which packages it is able to import as dependencies depends on whether it contains a project file. If it does not, it can import any root package. Otherwise, if it does contain a `Package.toml` file, it can only import those packages which are identified in its `[deps]` sections (by uuid and name).

Reviewer: Perhaps you were thinking above of what dependencies the packages in a package directory can load? If there's a project file in a package in a package directory, its deps section determines what it can load and what each dependency name means. Otherwise, package names are looked up at the top level like in the REPL or Main.

Author: I agree, I wanted to motivate the later sections, but the first part of this change doesn't read correctly.

**The roots map** is determined by enumerating the contents of the package directory to generate a list of all packages that exist.
Additionally, a UUID will be assigned to each entry as follows: For a given package found inside the folder `Name`...

> 1. If `Name/Project.toml` exists and has a `uuid` entry, then `uuid` is that value.

Reviewer: I think using the name `X` consistently throughout the document was clearer. Now we use `X` early on and then switch to `Name` later which seem more likely to be an actual package name and less clearly a placeholder.

Author: Yep, I'll switch back. Was just interesting in seeing whether it felt better or not. Either way, we would certainly need to be more consistent about the change if we did decide that a longer name (`Name`, `Xray`, `FooBar`, or whatever) was clearer (defined as simpler to read and more obvious that each instance of X must be the same string, and not just a unique arbitrary placeholder).

)
```

This example map includes three different kinds of package locations (the first and third are part of the default load path):

1. The private `Priv` package is "[vendored](https://stackoverflow.com/a/35109534/659248)" inside the `App` repository.
2. The public `Priv` and `Zebra` packages are in the system depot, where packages installed and managed by the system administrator live. These are available to all users on the system.
3. The `Pub` package is in the user depot, where packages installed by the user live. These are only available to the user who installed them.

### Package directories

> Package directories provide a simpler environment, but less powerful or controlled. The set of root packages available is defined simply as the set of files in the directory that "look like" packages. A package `X` is determined to exist if it finds one of the following files:

Reviewer: ```suggestion
Package directories provide a simpler environment, but less powerful or controlled. The set of root packages available is determined by the set of subdirectories that "look like" packages. A package `X` is determined to exist if one of the following files exists in the package directory:
```

>   - A file `X.jl`

Reviewer: ```suggestion
  - `X.jl`
```

>   - A file `X/src/X.jl`

Reviewer: ```suggestion
  - `X/src/X.jl`
```

>   - A file `X.jl/src/X.jl`

Reviewer: ```suggestion
  - `X.jl/src/X.jl`
```

> Additionally, which packages it is able to import as dependencies depends on whether that package contains a project file. If it does not, it can import any root package. Otherwise (if it does contain a `Package.toml` file), it can only import those packages which are identified in its `[deps]` sections (given by uuid and name).

Reviewer: ```suggestion
Additionally, which dependencies a package in a package directory is able to import depends on whether the package contains a project file:
* If it does not have a project file, it can import any root package—the same packages that can be loaded at the top level from `Main` and in the REPL.
* If it does have a project file, it can only import those packages which are identified in the `[deps]` section of the project file.

This dependency loading is part of the `graph` map for package directories, explained elsewhere.
```
